Worknet plugins/extensions

ngdentdev opened the pull request.

cloudmation committed:
wrote a basic worknet plugin

cloudmation committed:
also seek plugins in ~/.neo folder

cloudmation committed:
Merge branch 'main' into worknetPlugins

cloudmation committed:
added console out logging when plugins are loaded

cloudmation committed:
renamed sample logger and added more events to listen to

cloudmation committed:
added more info to console out onCommitting

cloudmation committed:
updated Extending NEO Worknet readme

devhawk commented on src/worknet-ext/logger/WorkNetLogger.cs:
```diff
@@ -0,0 +1,73 @@
+
+namespace WorkNetExt;
+
+public class WorkNetLogger : Plugin
```
I get this is a sample plugin, but is there some way we could do something that doesn't completely duplicate what WorkNet's built in log plugin does? Write to a file maybe?

devhawk commented on src/worknet-ext/logger/WorkNetLogger.cs:
```diff
@@ -0,0 +1,73 @@
+
+public class WorkNetLogger : Plugin
+{
+    NeoSystem? neoSystem;
```
Plugins can override the `Configure` method and call the `GetConfiguration` protected method to retrieve configuration information (in lieu of configuring in code). Here's an [example from the TokenTracker plugin](https://github.com/neo-project/neo-modules/blob/master/src/TokensTracker/TokensTracker.cs#L49): 

```cs
protected override void Configure()
{
    IConfigurationSection config = GetConfiguration();
    _dbPath = config.GetValue("DBPath", "TokensBalanceData");
    _shouldTrackHistory = config.GetValue("TrackHistory", true);
    _maxResults = config.GetValue("MaxResults", 1000u);
    _network = config.GetValue("Network", 860833102u);
    _enabledTrackers = config.GetSection("EnabledTrackers").GetChildren().Select(p => p.Value).ToArray();
}
```

Our sample plugin should demonstrate using Configure/GetConfiguration.

Note, `GetConfiguration` reads from the virtual `ConfigFile` property which (eventually) reads from the static `PluginsDirectory` property. The `PluginsDirectory` property - like the main plugin loaded - is relative to the assembly location. I suspect that worknet compatible plugins will need to overrride `ConfigFile` to read from a location that is relative to the two locations we are loading plugins from.

devhawk replied:
One other thing, [Neo 3.6](https://github.com/neo-project/neo/issues/2844) is likely to ship soon. It might be worth updating the logic of the platform `Plugin` class. In particular, making the platform `Plugin.LoadPlugins` method public + take an optional directory parameter. Then maybe we can also modify the `ConfigFile/RootPath/PluginsDirectory` properties to track the folder the plugin was originally loaded from.

devhawk commented on src/worknet/Node/PluginHandler.cs:
```diff
@@ -0,0 +1,46 @@
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Failed to load plugin: {type.FullName}", ex);
```
THe platform `LoadPlugin` logs an error on exception instead of throwing. Why did you change this?

ngdentdev replied:
I did that during debugging, replaced the code to log now

devhawk commented on src/worknet/Node/PluginHandler.cs:
```diff
@@ -0,0 +1,46 @@
+
+class PluginHandler
+{
+    public static void LoadPlugins(string directory, IConsole console)
```
Since IConsole is a type specific to the command line library WorkNet uses, we should pass this as a TextWriter instead. Maybe we make it optional instead? Both of these will make it easier to update platform logic and eliminate the need for this class entirely.

devhawk reviewed (changes requested):
I made a few comments with suggested changes

devhawk commented on src/worknet/Node/PluginHandler.cs:
```diff
@@ -0,0 +1,46 @@
+using Neo.Plugins;
+using McMaster.Extensions.CommandLineUtils;
+
+class PluginHandler
```
static class?

cloudmation committed:
-updated worknet plugin sample to read from config file and write output to a file
- updated pluginHandler method to take TextWriter instead of IConsole so we don't have to depend on console

cloudmation committed:
made plugin handler a static class

devhawk commented on src/worknet/Node/PluginHandler.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    private static void Log(TextWriter? writer, string message)
```
Instead of adding a Log method, you can use the [null conditional operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-). Everywhere you currently call Log, use this instead:

```cs
writer?.WriteLine("whatever the message is")
```

ngdentdev replied:
oh good, didn't know c# had this as well.

devhawk commented on src/worknet-ext/logger/WorkNetLogger.cs:
```diff
@@ -0,0 +1,90 @@
+        GC.SuppressFinalize(this);
+    }
+
+    // Overwrite Config file method to find the config.json from the same directory as the plugin dll directory
```
Please expand this comment so that plug in developers undertand that overriding ConfigFile is an absolute requirement for building plugins that are worknet compatible

devhawk reviewed (approved):
ready to merge once null conditional operator use + expanded ConfigFile overload comment are addressed

cloudmation committed:
- updated comments and readme
- use writer?.WriteLine pattern to avoid the null check in PluginHandler class

cloudmation committed:
updated readme to include configfile override

devhawk commented:
:shipit:
